Formula/gcc.rb

```diff
@@ -87,21 +84,25 @@ def install
       --with-zstd=#{Formula["zstd"].opt_prefix}
       --with-pkgversion=#{pkgversion}
       --with-bugurl=#{tap.issues_url}
+      --with-system-zlib
     ]
-    # libphobos is part of gdc
-    args << "--enable-libphobos" if Hardware::CPU.intel?
 
     if OS.mac?
       cpu = Hardware::CPU.arm? ? "aarch64" : "x86_64"
       args << "--build=#{cpu}-apple-darwin#{OS.kernel_version.major}"
-      args << "--with-system-zlib"
+
+      # libgcc_s.1.dylib links to `@rpath/libgcc_s.1.1.dylib`.
+      ENV.append "LDFLAGS", "-Wl,-rpath,@loader_path"
```

Comment:
We need this for `libgcc_s.1.dylib` to be able to find `libgcc_s.1.1.dylib`.

They're actually distinct libraries (presumably `libgcc_s.1.dylib` is provided for compatibility with pre-existing binaries):
```
❯ file libgcc_s.*.dylib
libgcc_s.1.1.dylib: Mach-O universal binary with 1 architecture: [x86_64:Mach-O 64-bit x86_64 dynamically linked shared library, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|WEAK_DEFINES|BINDS_TO_WEAK|NO_REEXPORTED_DYLIBS>]
libgcc_s.1.dylib:   Mach-O universal binary with 1 architecture: [x86_64:Mach-O 64-bit x86_64 dynamically linked shared library, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|WEAK_DEFINES>]
```

`libgcc_s.1.dylib` links with `libgcc_s.1.1.dylib` through an `@rpath` reference:
```
❯ otool -L libgcc_s.1.dylib
libgcc_s.1.dylib:
        /usr/local/opt/gcc@12/lib/gcc/12/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.1.0)
        @rpath/libgcc_s.1.1.dylib (compatibility version 1.0.0, current version 1.1.0, reexport)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
```
but it contains no `LC_RPATH` entries itself:
```
❯ otool -l libgcc_s.1.dylib | rg -q LC_RPATH || echo no rpaths
no rpaths
```

Reply:
This would normally not be an issue, since `gcc-12` passes the appropriate `-rpath` flags to the linker:
```
❯ echo 'int main(){}' | gcc-12 -xc -
❯ otool -l a.out | rg -A2 LC_RPATH
          cmd LC_RPATH
      cmdsize 32
         path @loader_path (offset 12)
--
          cmd LC_RPATH
      cmdsize 88
         path /usr/local/Cellar/gcc@12/12.1.0_1/lib/gcc/12/gcc/x86_64-apple-darwin21/12 (offset 12)
--
          cmd LC_RPATH
      cmdsize 64
         path /usr/local/Cellar/gcc@12/12.1.0_1/lib/gcc/12/gcc (offset 12)
--
          cmd LC_RPATH
      cmdsize 64
         path /usr/local/Cellar/gcc@12/12.1.0_1/lib/gcc/12 (offset 12)
```
However, these Cellar references will get invalidated at version/revision bumps.

Reply:
We should not worry about `libgcc_s.1.dylib` it's there only for backward compatibility (quite old stuff, actually). It is expected behavior that the executables have to provide the rpath, not the library itself.

Reply:
Yes, I'm also unsure how the build handles `LDFLAGS`. I don't think it really hurts, however. But we can remove it if we need to re-run this. Unless you think it's worth restarting CI?

Reply:
I'd rather we don't ship something with weird rpath adjustment, especially as it may have an impact on the bottled formulas built against it. But then, I know this takes so long to go through CI…

Do you think this will be mergeable after this run? I mean, there are still a lot of Linux failures.

Reply:
> I'd rather we don't ship something with weird rpath adjustment, especially as it may have an impact on the bottled formulas built against it. 

Well, that's the idea: existing bottles are unlikely to have the right `LC_RPATH` commands, which is why we do it here. For these bottles, shipping without this change could break them.

But, then again, I think given that we're rebuilding that links to some `libgcc_s`, we should be fine without it.

This is assuming they're linking now to `libgcc_s.1.1`, which I haven't checked. Those that link to `libgcc_s.1` are likely to be broken.

> But then, I know this takes so long to go through CI…

Yes, that too. Hopefully our changes here make it so that this is the last really painful major version bump 🤞.

> Do you think this will be mergeable after this run? I mean, there are still a lot of Linux failures. But I guess all the choices we have are bad in some way…

Possibly, depending on how much breakage we're willing to tolerate.

I do I have an open `brew` PR that might make things a lot easier, especially in CI, but I need reviews: https://github.com/Homebrew/brew/pull/13631

Reply:
Nothing should link against `libgcc_s.1`, anyway.

If you rerun CI, let's remove that LDFLAGS. Otherwise, if you merge after this run, I'll open a follow-up to do it.

Reply:
Going to rerun. Will drop this.

Reply:
Unfortunately, `julia` failed to build with the error
```
  ERROR: Unable to load dependent library /private/tmp/julia-20220802-62326-1frzjwy/julia-1.7.3/usr/lib/julia/libgcc_s.1.dylib
  Message:dlopen(/private/tmp/julia-20220802-62326-1frzjwy/julia-1.7.3/usr/lib/julia/libgcc_s.1.dylib, 0x000A): Library not loaded: @rpath/libgcc_s.1.1.dylib
    Referenced from: /usr/local/Cellar/gcc/12.1.0/lib/gcc/current/libgcc_s.1.dylib
    Reason: tried: '/private/tmp/julia-20220802-62326-1frzjwy/julia-1.7.3/usr/lib/libgcc_s.1.1.dylib' (no such file), '/private/tmp/julia-20220802-62326-1frzjwy/julia-1.7.3/usr/bin/../lib/libgcc_s.1.1.dylib' (no such file), '/usr/local/lib/libgcc_s.1.1.dylib' (no such file), '/usr/lib/libgcc_s.1.1.dylib' (no such file)
```
which is exactly what I was trying to avoid here. Thoughts, FX?

Reply:
Julia copies symlinks of all the libraries in its own directory, to avoid this problem. But it shouldn't link against `libgcc_s.1`, just the `1.1`. It's a formula with few users, so maybe we can ship this and I will look into a fix.

There are three places to fix, but it's a rather simple fix:
- https://github.com/JuliaLang/julia/blob/master/deps/csl.mk
- https://github.com/JuliaLang/julia/blob/master/base/Makefile
- https://github.com/JuliaLang/julia/blob/master/Make.inc

I'll open a PR with them.

Reply:
A quick-and-dirty hack that I think might work would be to symlink `libgcc_s.1.1` into Julia's `lib/julia` too. Not 100% sure that'll work, but I suspect it will.

Reply:
PR for Julia: https://github.com/JuliaLang/julia/pull/46240

Reply:
This line is now gone.